core/unit_test/tstParticleGridCommunication.hpp

```diff
@@ -0,0 +1,445 @@
+
+    // Make some data to migrate, one for each rank neighbor. This tests
+    // only for neighbors within one ghost shell.
+    int num_data = 27;
+    const int id_index = 0;
+    const int pos_index = 1;
+    using DataTypes = Cabana::MemberTypes<int, double[3]>;
+    Cabana::AoSoA<DataTypes, Kokkos::HostSpace> data_host( "host", num_data );
+
+    // Get mesh info.
+    auto dx = local_grid->globalGrid().globalMesh().cellSize( Cajita::Dim::I );
+    auto dy = local_grid->globalGrid().globalMesh().cellSize( Cajita::Dim::J );
+    auto dz = local_grid->globalGrid().globalMesh().cellSize( Cajita::Dim::K );
+    auto lo_x = local_mesh.lowCorner( Cajita::Own(), Cajita::Dim::I );
+    auto lo_y = local_mesh.lowCorner( Cajita::Own(), Cajita::Dim::J );
+    auto lo_z = local_mesh.lowCorner( Cajita::Own(), Cajita::Dim::K );
+    auto width_x = local_mesh.extent( Cajita::Own(), Cajita::Dim::I ) / 2.0;
+    auto width_y = local_mesh.extent( Cajita::Own(), Cajita::Dim::J ) / 2.0;
+    auto width_z = local_mesh.extent( Cajita::Own(), Cajita::Dim::K ) / 2.0;
+    auto center_x = width_x + lo_x;
+    auto center_y = width_y + lo_y;
+    auto center_z = width_z + lo_z;
+    auto shift_x = width_x - ( halo_width - 0.1 ) * dx;
+    auto shift_y = width_y - ( halo_width - 0.1 ) * dy;
+    auto shift_z = width_z - ( halo_width - 0.1 ) * dz;
+
+    // Fill the data. Add particles near the local domain boundary in each
+    // direction.
+    int my_rank = -1;
+    MPI_Comm_rank( MPI_COMM_WORLD, &my_rank );
+    auto id_host = Cabana::slice<id_index>( data_host );
+    auto pos_host = Cabana::slice<pos_index>( data_host );
+    int nr = 0;
+    for ( int k = -1; k < 2; ++k )
+        for ( int j = -1; j < 2; ++j )
+            for ( int i = -1; i < 2; ++i, ++nr )
+            {
+                id_host( nr ) = my_rank;
+                pos_host( nr, 0 ) = center_x + i * shift_x;
+                pos_host( nr, 1 ) = center_y + j * shift_y;
+                pos_host( nr, 2 ) = center_z + k * shift_z;
+            }
+    // Add a particle on rank zero to force some resizing for sends.
+    if ( my_rank == 0 )
+    {
+        num_data++;
+        data_host.resize( num_data );
+        auto id_host = Cabana::slice<id_index>( data_host );
+        id_host( num_data - 1 ) = my_rank;
+        auto pos_host = Cabana::slice<pos_index>( data_host );
+        pos_host( num_data - 1, 0 ) = center_x + shift_x;
+        pos_host( num_data - 1, 1 ) = center_y + shift_y;
+        pos_host( num_data - 1, 2 ) = center_z + shift_z;
+    }
+    int local_num_data = num_data;
+
+    Cabana::AoSoA<DataTypes, Kokkos::HostSpace> initial( "initial", num_data );
+    Cabana::deep_copy( initial, data_host );
+    auto pos_initial = Cabana::slice<pos_index>( initial );
+
+    // Copy to the device.
+    Cabana::AoSoA<DataTypes, TEST_MEMSPACE> data_src( "data_src", num_data );
+    Cabana::deep_copy( data_src, data_host );
+
+    if ( test_type == 0 )
```

Comment:
Did you anticipate adding other values for `test_type`?

Reply:
I added a slice variant of `gridGather` for a second `test_type`